### PR TITLE
Rewrite votedFor and add leader propagation queue

### PR DESCRIPTION
**Pull request by kjarosh**

### CODE REVIEW

File: src/main/kotlin/com/github/davenury/ucac/consensus/raft/infrastructure/RaftLeader.kt [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        this.id = id
+        this.elected = false
+    }
+
+    suspend fun elect(id: GlobalPeerId) = mutex.withLock {
+        this.id = id
+        this.elected = true
+
+        propagateChanges(id)
+    }
```
Comment (Rados13) [inline src/main/kotlin/com/github/davenury/ucac/consensus/raft/infrastructure/RaftLeader.kt:50]: This function doesn't handle situations when you are elected as a leader. In that situation, you will block yourself from sending http requests to yourself and don't schedule sending heartbeats. Which then leads to very strange

File: src/main/kotlin/com/github/davenury/ucac/consensus/raft/infrastructure/RaftLeader.kt [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    private fun propagateChanges(leaderId: GlobalPeerId) {
+        while (propagationRequests.isNotEmpty()) {
+            val req = propagationRequests.pop()
+            with(CoroutineScope(leaderRequestExecutorService)) {
+                launch(MDCContext()) {
+                    req.cf.complete(propagateChange(leaderId, req.change))
+                }
+            }
+        }
+    }
```
Comment (Rados13) [inline src/main/kotlin/com/github/davenury/ucac/consensus/raft/infrastructure/RaftLeader.kt:91]: Why does `propagateChange` call `propagateChanges` which calls another method `propagateChange`?

**Review by Rados13 — CHANGES_REQUESTED** [review]
I have two issues. First I am not sure about renaming `VotedFor` to `RaftLeader`, because `votedFor` is a term from paper. Secondly, these changes make tests pass once on three times.

File: src/main/kotlin/com/github/davenury/ucac/consensus/raft/infrastructure/RaftConsensusProtocolImpl.kt [hunk 3]
```diff
@@ -692,54 +681,31 @@ class RaftConsensusProtocolImpl(
...
-//              TODO: Change after queue
-            else -> {
-                result.complete(ChangeResult(ChangeResult.Status.TIMEOUT, "There should always be a leader"))
-                throw Exception("There should always be a leader")
+        val electedLeader: GlobalPeerId?
+        mutex.withLock {
+            electedLeader = leader.elected()
+        }
+
+        if (electedLeader == globalPeerId) {
```
Comment (Davenury) [inline src/main/kotlin/com/github/davenury/ucac/consensus/raft/infrastructure/RaftConsensusProtocolImpl.kt:697]: As far as I understand, you can synchronize on L694 but when execution reaches L697 the electedLeader variable might be outdated?

File: src/main/kotlin/com/github/davenury/ucac/consensus/raft/infrastructure/RaftLeader.kt [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+                    "as it is not elected yet, queueing: $change")
+        }
+        return cf
+    }
+
+    private fun propagateChanges(leaderId: GlobalPeerId) {
+        while (propagationRequests.isNotEmpty()) {
+            val req = propagationRequests.pop()
+            with(CoroutineScope(leaderRequestExecutorService)) {
+                launch(MDCContext()) {
```
Comment (Davenury) [inline src/main/kotlin/com/github/davenury/ucac/consensus/raft/infrastructure/RaftLeader.kt:86]: why do you launch another MDCContext? shouldn't we use the one we have (somehow)?

File: src/main/kotlin/com/github/davenury/ucac/consensus/raft/infrastructure/RaftLeader.kt [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+                }
+            }
+        }
+    }
+
+    private suspend fun propagateChange(
+        leaderId: GlobalPeerId,
+        change: Change,
+    ): ChangeResult {
+        logger.info("Propagating change to leader ($leaderId): $change")
```
Comment (Davenury) [inline src/main/kotlin/com/github/davenury/ucac/consensus/raft/infrastructure/RaftLeader.kt:97]: I don't think its intuitive that class named `RaftLeader` logs `propagating change to leader` which means that this is not raft leader.

File: src/main/kotlin/com/github/davenury/ucac/consensus/raft/infrastructure/RaftLeader.kt [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        val leaderAddress = peerResolver.resolve(leaderId)
+        return try {
+            val url = "http://${leaderAddress.address}/consensus/request_apply_change"
+            val response = httpClient.post<ChangeResult>(url) {
+                contentType(ContentType.Application.Json)
+                accept(ContentType.Application.Json)
+                body = change
+            }
+            logger.info("Response from leader: $response")
+            response
```
Comment (Davenury) [inline src/main/kotlin/com/github/davenury/ucac/consensus/raft/infrastructure/RaftLeader.kt:107]: I don't know if I'm right, but I'd advise not to call specific httpClient from here, which can make testing hard.

Comment (Rados13) [reply src/main/kotlin/com/github/davenury/ucac/consensus/raft/infrastructure/RaftLeader.kt:107]: We use specific httpClient from each module :upside_down_face:

**Review by Rados13 — COMMENTED** [review]
This PR also looks good, maybe it's even more idiomatic for OOP. But I would also run it 50 times to check if there is no race condition.

**Pull request closed without merging**